libethereum/ChainParams.cpp

```diff
@@ -34,60 +34,107 @@ using namespace dev;
-	precompiled.insert(make_pair(Address(4), PrecompiledContract(15, 3, PrecompiledRegistrar::executor("identity"))));
+	loadConfig(_json, _stateRoot);
 }
 
-ChainParams::ChainParams(std::string const& _json, h256 const& _stateRoot)
+ChainParams::ChainParams(ChainParams const& _org)
 {
+	sealEngineName = _org.sealEngineName;
+	accountStartNonce = _org.accountStartNonce;
+	maximumExtraDataSize = _org.maximumExtraDataSize;
+	tieBreakingGas = _org.tieBreakingGas;
+	blockReward = _org.blockReward;
+	for (auto i: _org.otherParams)
+		otherParams[i.first] = i.second;
+	precompiled = _org.precompiled;
+	genesisState = _org.genesisState;
+	stateRoot = _org.stateRoot;
+	parentHash = _org.parentHash;
+	author = _org.author;
+	difficulty = _org.difficulty;
+	gasLimit = _org.gasLimit;
+	gasUsed = _org.gasUsed;
+	timestamp = _org.timestamp;
+	extraData = _org.extraData;
+	sealFields = _org.sealFields;
+	sealRLP = _org.sealRLP;
+}
+
+ChainParams ChainParams::setGenesis(string const& _json, h256 const& _stateRoot, ChainParams const& _org) const
+{
+	ChainParams cp(_org);
+
 	js::mValue val;
 	json_spirit::read_string(_json, val);
-	js::mObject obj = val.get_obj();
-	sealEngineName = obj["sealEngine"].get_str();
-	js::mObject genesis = obj["genesis"].get_obj();
-	js::mObject params = obj["params"].get_obj();
-
-	// params
+	js::mObject genesis = val.get_obj();
+
+	cp.parentHash = h256(genesis["parentHash"].get_str());
+	cp.author = genesis.count("coinbase") ? h160(genesis["coinbase"].get_str()) : h160(genesis["author"].get_str());
+	cp.difficulty = genesis.count("difficulty") ? u256(fromBigEndian<u256>(fromHex(genesis["difficulty"].get_str()))) : 0;
+	cp.gasLimit = u256(fromBigEndian<u256>(fromHex(genesis["gasLimit"].get_str())));
+	cp.gasUsed = genesis.count("gasUsed") ? u256(fromBigEndian<u256>(fromHex(genesis["gasUsed"].get_str()))) : 0;
+	cp.timestamp = u256(fromBigEndian<u256>(fromHex(genesis["timestamp"].get_str())));
+	cp.extraData = bytes(fromHex(genesis["extraData"].get_str()));
+
+	// magic code for handling ethash stuff:
+	if ((genesis.count("mixhash") || genesis.count("mixHash")) && genesis.count("nonce"))
 	{
-		accountStartNonce = u256(fromBigEndian<u256>(fromHex(params["accountStartNonce"].get_str())));
-		maximumExtraDataSize = u256(fromBigEndian<u256>(fromHex(params["maximumExtraDataSize"].get_str())));
-		tieBreakingGas = params.count("tieBreakingGas") ? params["tieBreakingGas"].get_bool() : true;
-		blockReward = u256(fromBigEndian<u256>(fromHex(params["blockReward"].get_str())));
-		for (auto i: params)
-			if (i.first != "accountStartNonce" && i.first != "maximumExtraDataSize" && i.first != "blockReward" && i.first != "tieBreakingGas")
-				otherParams[i.first] = i.second.get_str();
+		h256 mixHash(genesis[genesis.count("mixhash") ? "mixhash" : "mixHash"].get_str());
+		h64 nonce(genesis["nonce"].get_str());
+		cp.sealFields = 2;
+		cp.sealRLP = rlp(mixHash) + rlp(nonce);
 	}
+	cp.stateRoot = _stateRoot ? _stateRoot : cp.calculateStateRoot();
```

Comment:
Hm, we have not yet set / not modified the genesis state here. This should be moved to after the call of setGenesisState or inside it.